GH1931: Add ability to set NuGet.config through env var or cake.config

Cheesebaron opened the pull request:
Implements #1931 

With these changes you can now explicitly set the path for the NuGet.config file either through an env var or cake.config.

This should work for both the In Process client and the one invoking NuGet.exe.

dnfclas commented:
@Cheesebaron,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

patriksvensson commented on src/Cake.NuGet/Cake.NuGet.csproj:
```diff
@@ -26,6 +26,7 @@
     <PackageReference Include="NuGet.Frameworks" Version="4.3.0" />
     <PackageReference Include="NuGet.Versioning" Version="4.3.0" />
     <PackageReference Include="NuGet.ProjectModel" Version="4.3.0" />
+    <PackageReference Include="System.ValueTuple" Version="4.4.0" />
```
We don't want to take on this dependency at this point.

Cheesebaron replied:
Alright, will change this 👍

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -165,5 +168,26 @@ private static NuGetVersion GetNuGetVersion(PackageReference package, NuGetSourc
             return null;
         }
+
+        private static (string root, string fileName) GetNuGetConfigPath(ICakeEnvironment environment, ICakeConfiguration config, ILogger log)
```
Could be a `Tuple<DirectoryPath, FilePath>` here.

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -22,6 +22,7 @@
 using NuGet.Versioning;
 using PackageReference = Cake.Core.Packaging.PackageReference;
 using PackageType = Cake.Core.Packaging.PackageType;
+using Path = System.IO.Path;
```
Don't use `System.IO`. Use available classes in `Cake.Core.IO`.

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -165,5 +168,26 @@ private static NuGetVersion GetNuGetVersion(PackageReference package, NuGetSourc
+            string configFileName;
+            if (!string.IsNullOrEmpty(nugetConfigFile))
+            {
+                rootPath = Path.GetDirectoryName(nugetConfigFile);
```
Use `DirectoryPath` here and get the name from that one.

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -165,5 +168,26 @@ private static NuGetVersion GetNuGetVersion(PackageReference package, NuGetSourc
+            if (!string.IsNullOrEmpty(nugetConfigFile))
+            {
+                rootPath = Path.GetDirectoryName(nugetConfigFile);
+                configFileName = Path.GetFileName(nugetConfigFile);
```
Use `FilePath` here and get the name from that one.

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -165,5 +168,26 @@ private static NuGetVersion GetNuGetVersion(PackageReference package, NuGetSourc
+                configFileName = null;
+            }
+
+            log.LogDebug($"Found NuGet.config at: {rootPath}{Path.DirectorySeparatorChar}{configFileName}");
```
Move logging to `NuGetPackageInstaller` instead.

Cheesebaron replied:
Move it where? We are in NuGetPackageInstall.cs here. Or do you mean into the ctor of it?

patriksvensson commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -165,5 +168,26 @@ private static NuGetVersion GetNuGetVersion(PackageReference package, NuGetSourc
+
+            log.LogDebug($"Found NuGet.config at: {rootPath}{Path.DirectorySeparatorChar}{configFileName}");
+
+            return (rootPath, configFileName);
```
Return `Tuple<DirectoryPath, FilePath>` here instead.

patriksvensson commented on src/Cake.NuGet/NuGetPackageInstaller.cs:
```diff
@@ -240,6 +240,14 @@ private FilePath GetNuGetPath()
 
+            // Config
+            var nugetConfig = config.GetValue(Constants.NuGet.ConfigFile);
+            if (!string.IsNullOrWhiteSpace(nugetConfig))
```
I would want to see a unit tests for this.

Cheesebaron commented:
@patriksvensson I've addressed your comments in your review. As for the tests, do you have ideas for more tests? I don't see any tests for stuff in the `Cake.Nuget.Install` namespace as I also pointed out to @mholo65 on gitter.

mholo65 commented on src/Cake.NuGet/Install/NuGetPackageInstaller.cs:
```diff
@@ -59,9 +59,17 @@ internal sealed class NuGetPackageInstaller : INuGetPackageInstaller
+            var nugetConfigDirectoryPath = nugetConfig.Item1;
+            var nugetConfigFilePath = nugetConfig.Item2;
+
+            _nugetLogger.LogDebug($"Found NuGet.config at: {nugetConfigFilePath}");
```
Use `_log` Instead. `_nugetLogger` is just for NuGet.Client internals.

Cheesebaron replied:
Alright 👍 Done

Cheesebaron commented:
Rebased on latest develop

devlead commented:
@Cheesebaron your changes have been merged, thanks for your contribution 👍
